File: OS/Page Replacement/LRU.c
#include <stdio.h>
#include <stdlib.h>

void main()
{
    int noFrames,noPages, frames[10], pages[20], timestamp[10], fault=0, hit=0;
    printf("Enter number of frames : ");
    scanf("%d", &noFrames);
    for (int i=0; i<noFrames; i++)
    {
        frames[i] = -1;
    }
    for (int i=0; i<noFrames; i++)
    {
        timestamp[i] = -1;
    }
    printf("Enter total number of pages : ");
    scanf("%d", &noPages);
    printf("Enter sequence : \n");
    for (int i=0; i<noPages; i++)
    {
        scanf("%d", &pages[i]);
    }
    
    for (int i=0; i<noPages; i++)
    {
        int count = 0;
        for (int j=0; j<noFrames; j++)
        {
            if (frames[j] != pages[i])
            {
                count++;
            }
            else
            {
                hit++;
                timestamp[j] = i;
                break;
            }
        }
        if(i<noFrames && count == noFrames)
        {
            frames[fault] = pages[i];
            timestamp[fault] = i;
            fault++;
        }
        else if(count == noFrames)
        {
            fault++;
            int min=99, minElement;
            for (int j=0; j<noFrames; j++)
            {
                if(timestamp[j]<min)
                {
                    min = timestamp[j];
                    minElement = j;
                }
            }
            frames[minElement] = pages[i];
            timestamp[minElement] = i;
        }
        printf("Page : %d\tFrames : ", pages[i]);
        for (int j=0; j<noFrames; j++)
        {
            printf("%d\t", frames[j]);
        }
        printf("\n");
    }
    printf("\nPage Faults : %d", fault);
    printf("\nNumber of Hits : %d",hit);
    printf("\nNumber of Miss : %d",noPages - hit);
} 

File: POA/Page Replacement/py/LRU.py
frameno = int(input("Enter total number of frames: "))
pageno = int(input("Enter total number of pages: "))
pages = []
frames = [-1] * frameno
hit = 0
miss = 0

for i in range(pageno):
    pages.append(int(input("Enter page: ")))

print("Pages:", pages)

def find_lru(i):
        point=[-1]*frameno
        for j in range(frameno):
            for k in range(i):
                if(frames[j]==pages[k]):
                    point[j]=k
        lru = pages[min(point)]
        return lru

def print_frame():
    for k in range(frameno):
        if frames[k] == -1:
            print("-\t", end="")
        else:
            print("%d\t" % frames[k], end="")
    print()

for i in range(pageno):
        if pages[i] in frames:
            hit += 1
            print_frame()
        else:
            if pages[i] not in frames:
                miss += 1
                for j in range(frameno):
                    if frames[j] == -1:
                        frames[j] = pages[i]
                        break
                else:
                    lru = find_lru(i)
                    frames[frames.index(lru)] = pages[i]
                    
                print_frame()
            else:
                print_frame()

print("Number of Page Faults:%d" % miss)
print("Number of Page Hits:%d" % hit)
re2=hit/(miss+hit) 
print(re2*100)
